### PR TITLE
depends: cleanup package configure flags

### PR DESCRIPTION
**Pull request by fanquake**
Related to #16354.

This PR adds additional configure flags to packages in depends to explicitly disable features we aren't using; similar to #16183. It also fixes passing `--without-tools` to `qrencode`.

I've added `--disable-drafts` to `zeromq`:
```bash
Build and install draft classes and methods [default=yes]
```

I'm not entirely sure how far we want to take this. i.e in the `zeromq` package we explicitly pass `--without-libsodium`, even though it's disabled by default.

Do we also want to explicitly pass all the other `--without` flags? :
```bash
  --with-libgssapi_krb5   require libzmq build with libgssapi_krb5
                          [default=no]
  --with-libsodium        use libsodium instead of built-in tweetnacl
                          [default=no]
  --with-pgm              build libzmq with PGM extension. Requires pkg-config
                          [default=no]
  --with-norm             build libzmq with NORM protocol extension,
                          optionally specifying norm path [default=no]
  --with-vmci             build libzmq with VMCI transport [default=no]
```

### CODE REVIEW

**Comment by Sjors** [comment]
Less is more :-) I'm able to build and run QT on macOS 10.14.5 with 9a3e237. `test/functional/interface_zmq.py` passes.

Explicitly passing all `--without` flags seems like a good idea; it means we don't to worry about the upstream default changing and it's also easier to reason about.

**Comment by dongcarl** [comment]
Concept ACK! Looking at the DrahtBot logs, `xproto` and `xextproto` are both headers-only, so let's remove `--disable-shared` from them!

**Comment by practicalswift** [comment]
Concept ACK

Compilation time before and after number would be interesting.

Nice cleanup regardless of compilation time impact though.

**Comment by fanquake** [comment]
Rebased for #16408.

File: depends/packages/zeromq.mk [hunk 1]
```diff
@@ -6,7 +6,8 @@ $(package)_sha256_hash=bcbabe1e2c7d0eec4ed612e10b94b112dd5f06fcefa994a0c79a45d83
 $(package)_patches=0001-fix-build-with-older-mingw64.patch 0002-disable-pthread_set_name_np.patch
 
 define $(package)_set_vars
-  $(package)_config_opts=--without-docs --disable-shared --without-libsodium --disable-curve --disable-curve-keygen --disable-perf --disable-Werror
+  $(package)_config_opts=--without-docs --disable-shared --disable-curve --disable-curve-keygen --disable-perf --disable-Werror --disable-drafts
```
Comment (dongcarl) [inline depends/packages/zeromq.mk]: We can add: `--disable-libunwind` `--disable-radix-tree` `--without-gcov` `--without-militant`

File: depends/packages/libXau.mk [hunk 2]
```diff
@@ -6,7 +6,7 @@ $(package)_sha256_hash=fdd477320aeb5cdd67272838722d6b7d544887dfe7de46e1e7cc0c27c
 $(package)_dependencies=xproto
 
 define $(package)_set_vars
-  $(package)_config_opts=--disable-shared
+  $(package)_config_opts=--disable-shared --disable-lint-library --without-lint
```
Comment (dongcarl) [inline depends/packages/libXau.mk]: Nit: Took a look at `--disable-xthreads` in the autoconf files... It seems that it's currently enabled. I'm not sure what the impact is of disabling it so we'll leave it for now.

File: depends/packages/libxcb.mk [hunk 3]
```diff
@@ -6,7 +6,7 @@ $(package)_sha256_hash=98d9ab05b636dd088603b64229dd1ab2d2cc02ab807892e107d674f9c
 $(package)_dependencies=xcb_proto libXau
 
 define $(package)_set_vars
-$(package)_config_opts=--disable-static
+$(package)_config_opts=--disable-static --disable-build-docs --without-doxygen
```
Comment (dongcarl) [inline depends/packages/libxcb.mk]: This is a candidate for another issue... It seems that we build all of the default XCB Extensions, but the `-qt-xcb` option we supply to QT might cause it to build them as well. (source: https://doc.qt.io/qt-5.9/linux-requirements.html)

File: depends/packages/xproto.mk [hunk 4]
```diff
@@ -5,7 +5,7 @@ $(package)_file_name=$(package)-$($(package)_version).tar.bz2
 $(package)_sha256_hash=636162c1759805a5a0114a369dffdeccb8af8c859ef6e1445f26a4e6e046514f
 
 define $(package)_set_vars
-$(package)_config_opts=--disable-shared
+$(package)_config_opts=--without-fop --without-xmlto --without-xsltproc
```
Comment (dongcarl) [inline depends/packages/xproto.mk]: Nit: Looking at https://gist.github.com/dongcarl/892d7deda8da3ff72317009ecbb07ffb#file-xproto-configure-help-L80-L90, I'm not sure what the specs and prototypes are... If anyone else knows please say so, otherwise we can leave it.

File: depends/packages/zeromq.mk [hunk 5]
```diff
@@ -6,7 +6,8 @@ $(package)_sha256_hash=bcbabe1e2c7d0eec4ed612e10b94b112dd5f06fcefa994a0c79a45d83
...
+  $(package)_config_opts += --without-libsodium --without-libgssapi_krb5 --without-pgm --without-norm --without-vmci
```
Comment (dongcarl) [inline depends/packages/zeromq.mk]: Looking at https://gist.github.com/dongcarl/892d7deda8da3ff72317009ecbb07ffb#file-zeromq-configure-help-L114-L120, I'm reminded that after we're able to use Guix for release builds, we will finally be able to, and definitely should specify what the minimal kernel headers are that we target for each architecture. We don't want to target kernel headers that are too new, have the zeromq build auto-assume we have the fancy new `ekquelect` (not an actual thing) available, and then not be able to run on older kernels.

**Review by dongcarl — COMMENTED** [review]
Much needed work, many thanks for starting it!

I reviewed this PR with the help of [this gist](https://gist.github.com/dongcarl/892d7deda8da3ff72317009ecbb07ffb), which contains all the `configure --help` outputs of the packages you've touched here.

**Comment by dongcarl** [comment]
Thinking a little harder about our policy w/re configure flags here... I believe that our approach with these cleanups should be to explicitly disable as many features and packages as possible without changing Bitcoin Core's functionality.

Going forward, we can keep our builds sane by simply requiring (at review time) a diff of the `configure --help` output for every package version bump. We will change configure flags only to support new Bitcoin Core features or to disable new and unneeded features/packages for depends package version bumps.

**Comment by theuni** [comment]
@dongcarl I agree with the spirit of that idea, but I don't think we should actually attempt to automate or codify.

We need to be a little more surgical about this, imo. Remember that buildsystems are imperfect. An explicit "--without-foo" may accidentally produce a different result from an auto-detected missing "foo". The most obvious example of this would be configuring Core with an explicit CXXFLAGS: "./configure CXXFLAGS="-O2 -g". Being explicit there means clobbering the warnings and hardening that would've been added otherwise.

tl;dr: Concept ACK, but let's not go around flipping _every_ switch we can find indiscriminately. Will review.

File: depends/packages/xproto.mk (hunk 4 above)
Comment (fanquake) [reply depends/packages/xproto.mk]: From the looks of this [mailing list post](https://lists.x.org/archives/xorg-announce/2010-February/001249.html) they are probably related to documentation generation?

> Document building enablement macros. This allows a builder to skip
document generation if such documents are not desired. They should not
be used to workaround tooling issues, use macros above. By default,
building is enabled. The default value can be changed from configure.ac
by invoking the macro like XORG_ENABLE_SPECS(no)


> XORG_ENABLE_DOCS            ENABLE_DOCS        General documentation
> XORG_ENABLE_DEVEL_DOCS      ENABLE_DEVEL_DOCS  More specifically developer's documentation
> XORG_ENABLE_SPECS           ENABLE_SPECS       More specifically functional specifications

File: depends/packages/libxcb.mk (hunk 3 above)
Comment (fanquake) [reply depends/packages/libxcb.mk]: Opened an issue to track that: https://github.com/bitcoin/bitcoin/issues/16447

File: depends/packages/zeromq.mk (hunk 5 above)
Comment (fanquake) [reply depends/packages/zeromq.mk]: Is this something you'd like to document here, or should we start tracking future Guix related improvements somewhere else? Depends how much there is to keep track of.

**Comment by fanquake** [comment]
> tl;dr: Concept ACK, but let's not go around flipping every switch we can find indiscriminately. Will review.

Fair call. I've made some of the changes suggested by Carl above, and left some comments inline. I'll leave this until you've done a first pass review.

> Showing a diff of configure --help as part of a depends PR couldn't possibly be a bad thing!

I think configure diffs are a nice idea, and at least _they_ can be automated.

File: depends/packages/libXau.mk (hunk 2 above)
Comment (dongcarl) [reply depends/packages/libXau.mk]: Yeah a comment would be great for the future!

File: depends/packages/zeromq.mk (hunk 5 above)
Comment (dongcarl) [reply depends/packages/zeromq.mk]: Tracked https://github.com/bitcoin/bitcoin/issues/16496

File: depends/packages/xproto.mk (hunk 4 above)
Comment (dongcarl) [reply depends/packages/xproto.mk]: Let's just do `--disable-specs`, you can put a comment there for the other flags or don't, either way is fine :-)

File: depends/packages/libXau.mk (hunk 2 above)
Comment (fanquake) [reply depends/packages/libXau.mk]: Added a comment.

File: depends/packages/xproto.mk (hunk 4 above)
Comment (fanquake) [reply depends/packages/xproto.mk]: Updated with `--disable-specs`.

**Review by theuni — COMMENTED** [review]
Left a few things to check.

The ```--with-militant``` is the type of switch-flipping that makes me really nervous. Can we make sure that binary results don't differ before/after these changes? IMO Anything that changes the bitcoind binary should be reviewed separately.

File: depends/packages/libxcb.mk (hunk 3 above)
[the remark this answers is not in the document]
Comment (fanquake) [reply depends/packages/libxcb.mk]: Updated with `--without-launchd`. Docs are:
```bash
--with-launchd Build with support for Apple's launchd (default: auto)
```

File: depends/packages/qrencode.mk [hunk 6]
```diff
@@ -5,7 +5,7 @@ $(package)_file_name=$(package)-$($(package)_version).tar.bz2
 $(package)_sha256_hash=efe5188b1ddbcbf98763b819b146be6a90481aac30cfc8d858ab78a19cde1fa5
 
 define $(package)_set_vars
-$(package)_config_opts=--disable-shared -without-tools --disable-sdltest
+$(package)_config_opts=--disable-shared --without-tools --without-tests --disable-sdltest
```
[the remark this answers is not in the document]
Comment (fanquake) [reply depends/packages/qrencode.mk]: Updated to remove `--without-libiconv-prefix`. Docs:
```bash
--with-libiconv-prefix[=DIR]  search for libiconv in DIR/include and DIR/lib
```

**Review by fanquake — COMMENTED** [review]
I've addressed two of @theuni's comments (libxcb and qrencode).

For the remaining two, I'm adding some [changes](https://github.com/bitcoin-core/bitcoin-maintainer-tools/pull/23) to our [`build-for-compare.py`](https://github.com/bitcoin-core/bitcoin-maintainer-tools/blob/master/build-for-compare.py) script so that you can pass a depends prefix through to configure. That should make it easier to do binary comparisons of depends changes. I'll follow up with some comparisons of the `--with-militant` (zmq) changes shortly.

> Doesn't this install generated spec files? If so, does libXau not need them?

Still need to figure out and address this.

**Comment by laanwj** [comment]
Concept ACK

**Comment by fanquake** [comment]
I've opened [1 more PR](https://github.com/bitcoin-core/bitcoin-maintainer-tools/pull/24) to the maintainer-tools repo that ensures we get deterministic `bitcoin-qt` builds.

> Disabling this (the current default) sounds like a safety downgrade...

I agree, and have now removed that change.

Passing `--with-militant` to zmqs `./configure` (the default) defines `ZMQ_ACT_MILITANT`. Looking at the zeromq source, the usages of `ZMQ_ACT_MILITANT` are all in [options.cpp](https://github.com/zeromq/libzmq/blob/master/src/options.cpp). 
I've included some of the usage below:
```cpp
static int sockopt_invalid ()
{
#if defined(ZMQ_ACT_MILITANT)
    zmq_assert (false);
#endif
    errno = EINVAL;
    return -1;
}
```

```cpp
        default:
#if defined(ZMQ_ACT_MILITANT)
            //  There are valid scenarios for probing with unknown socket option
            //  values, e.g. to check if security is enabled or not. This will not
            //  provoke a militant assert. However, passing bad values to a valid
            //  socket option will, if ZMQ_ACT_MILITANT is defined.
            malformed = false;
#endif
            break;
    }
```

```cpp
#if defined(ZMQ_ACT_MILITANT)
    //  There is no valid use case for passing an error back to the application
    //  when it sent malformed arguments to a socket option. Use ./configure
    //  --with-militant to enable this checking.
    if (malformed)
        zmq_assert (false);
#endif
    errno = EINVAL;
    return -1;
}
```

FWIW the diff of the bitcoinds from a comparison build, one which had `--without-militant` passed to zeromq, is very large.

**Comment by fanquake** [comment]
Rebased for #16533.

**Comment by dongcarl** [comment]
@fanquake So the patchset as it stands right now does not change add the `--with-militant` flag, and thus the `build-for-compare.py` results are identical? I think this is merge-able, and we can deal with `--with-militant` in another PR/issue.

**Comment by fanquake** [comment]
@dongcarl Yes I'm seeing identical results. 

I've just done another [build-for-compare](https://github.com/bitcoin-core/bitcoin-maintainer-tools/blob/master/build-for-compare.py) using the following (inside a Buster Docker container):
```bash
git fetch upstream pull/16370/head:16370 && git checkout 16370

# build depends for c295cba5a2f934e51a7c8610ab4c58b8e9d56619
cd depends
RAPIDCHECK=1 make -j5
cd ..

# build for compare into /tmp/c295cba5a2f934e51a7c8610ab4c58b8e9d56619
../bitcoin-maintainer-tools/build-for-compare.py c295cba5a2f934e51a7c8610ab4c58b8e9d56619  --prefix=/bitcoin/depends/x86_64-pc-linux-gnu --executables=src/bitcoind,src/qt/bitcoin-qt --parallelism=5 --tgtdir=/tmp/c295cba5a2f934e51a7c8610ab4c58b8e9d56619

# jump back to last commit in the branch before the changes in this PR
git checkout 24f29790cd7e1001c0f923130ea24b18d378bbb5

# re build depends
cd depends
RAPIDCHECK=1 make -j5
cd ..

# build for compare into /tmp/24f29790cd7e1001c0f923130ea24b18d378bbb5
../bitcoin-maintainer-tools/build-for-compare.py 24f29790cd7e1001c0f923130ea24b18d378bbb5  --prefix=/bitcoin/depends/x86_64-pc-linux-gnu --executables=src/bitcoind,src/qt/bitcoin-qt --parallelism=5 --tgtdir=/tmp/24f29790cd7e1001c0f923130ea24b18d378bbb5

# compare
shasum /tmp/24f29790cd7e1001c0f923130ea24b18d378bbb5/*.stripped /tmp/c295cba5a2f934e51a7c8610ab4c58b8e9d56619/*.stripped
a97c2d0626fa5b115b32c37604703aacc577845a  /tmp/24f29790cd7e1001c0f923130ea24b18d378bbb5/bitcoin-qt.24f29790cd7e1001c0f923130ea24b18d378bbb5.stripped
7c78eb1760cf901f6dad7299ef6a270d73f3d52a  /tmp/24f29790cd7e1001c0f923130ea24b18d378bbb5/bitcoind.24f29790cd7e1001c0f923130ea24b18d378bbb5.stripped
a97c2d0626fa5b115b32c37604703aacc577845a  /tmp/c295cba5a2f934e51a7c8610ab4c58b8e9d56619/bitcoin-qt.c295cba5a2f934e51a7c8610ab4c58b8e9d56619.stripped
7c78eb1760cf901f6dad7299ef6a270d73f3d52a  /tmp/c295cba5a2f934e51a7c8610ab4c58b8e9d56619/bitcoind.c295cba5a2f934e51a7c8610ab4c58b8e9d56619.stripped

diffoscope 24f29790cd7e1001c0f923130ea24b18d378bbb5/bitcoin-qt.24f29790cd7e1001c0f923130ea24b18d378bbb5.stripped c295cba5a2f934e51a7c8610ab4c58b8e9d56619/bitcoin-qt.c295cba5a2f934e51a7c8610ab4c58b8e9d56619.stripped

```

**Comment by dongcarl** [comment]
ACK c295cba5a2f934e51a7c8610ab4c58b8e9d56619

**Comment by Sjors** [comment]
c295cba still builds for me on macOS.

**Comment by meshcollider** [comment]
Looks good to me, since dongcarl and theuni have both reviewed and the gitian build looks fine
